### PR TITLE
prepare actions to include qc report

### PR DESCRIPTION
**Pull request by zsigmas**
- Compacts GH actions building process in a single file equal, with minor changes, to the one we had before
- Includes QC report steps in GH actions
- We lose parallel execution of jobs, (execution time is low anyways and most time of parallel jobs is consumed in pod start up anyway)
- Parallel loss is outweighed, in my opinion, by a simplified GH actions flow
- Because the same pod is used, all artifacts generated are available across steps this simplifes including the qc report

The build will fail as actions points to @main and those actions do not exist there yet. They will build once the PR merged

### CODE REVIEW

File: .github/workflows/check.yml
Comment (sorinvoicu) [inline .github/workflows/check.yml:1]: why merge check/lintr/pkgdown ?

File: .github/workflows/package_building.yml [hunk 1]
```diff
@@ -16,43 +16,52 @@ on:
...
         shell: Rscript {0}
-
+  
       - name: Check URLs 🌐
         run: |
           lychee . --format markdown --verbose --no-progress >> $GITHUB_STEP_SUMMARY
         shell: bash  
 
-      - name: Checkout gh-pages branch ⬇️
-        if: github.ref_name == 'main'
```
Comment (sorinvoicu) [inline .github/workflows/package_building.yml:46]: why generate pkgdown docs for every commit?

File: workflows.md [hunk 2]
```diff
@@ -48,24 +48,24 @@ on:
...
     name: Lintr 🔍
-    uses: boehringer-ingelheim/dv.templates/.github/workflows/lintr.yml@main
+    uses: boehringer-ingelheim/dv.templates/.github/workflows/lintr.yml@local
     with:
       lintr_error_on_lint: true
 
   gitleaks:
     name: Gitleaks 🌧️
-    uses: boehringer-ingelheim/dv.templates/.github/workflows/gitleaks.yml@main
+    uses: boehringer-ingelheim/dv.templates/.github/workflows/gitleaks.yml@local
```
Comment (sorinvoicu) [inline workflows.md:61]: this should still be @main before merging

File: .github/workflows/package_building.yml [hunk 3]
```diff
@@ -16,43 +16,52 @@ on:
...
       image: ${{ vars.CI_IMAGE }}
+    env:
+      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
 
     steps:
       - name: Checkout project ⬇️
         uses: actions/checkout@v4
 
-      - name: Install package dependencies 📄
-        uses: boehringer-ingelheim/dv.templates/.github/actions/dependencies@main
```
Comment (sorinvoicu) [inline .github/workflows/package_building.yml:34]: why is this removed?

File: pkg_pipeline/single_build.R [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+pkg_name <- read.dcf("DESCRIPTION")[1, "Package"]
+pkg_version <- read.dcf("DESCRIPTION")[1, "Version"]
+
+# Building ----
+
+message("############################")
+message("###### INSTALLING (S) ######")
+message("############################")
+
+pak::local_install_deps(".", upgrade=FALSE, ask=FALSE, dependencies = TRUE)
```
Comment (sorinvoicu) [inline pkg_pipeline/single_build.R:20]: I would say keep the deps step removed above (since it's a dependency in all steps, not just check/lintr/pkgdown) and comment these 2 lines

Comment (sorinvoicu) [reply pkg_pipeline/single_build.R:20]: also why not add the qc report code in this repo in the first place? it uses nobody outside BI to have it in the package repo

**Review by sorinvoicu — CHANGES_REQUESTED** [review]

File: pkg_pipeline/single_build.R [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+message("###### INSTALLING (F) ######")
+message("############################")
+
+# Checking ----
+
+message("##########################")
+message("###### CHECKING (S) ######")
+message("##########################")
+
+check_results <- rcmdcheck::rcmdcheck(error_on = "never", args = "--no-tests")
```
Comment (sorinvoicu) [inline pkg_pipeline/single_build.R:33]: would definitely not serialize steps in a big file. KISS, remember?

**Review by sorinvoicu — CHANGES_REQUESTED** [review]

File: pkg_pipeline/single_build.R (hunk 5 above)
Comment (sorinvoicu) [reply pkg_pipeline/single_build.R:33]: + why wait for fails if you can sees them all at once and hopefully fix them in one go?

File: .github/workflows/check.yml
Comment (zsigmas) [reply .github/workflows/check.yml:1]: Main point is that pkgdown uses the output of tests to generate the quality control documentation, it is simpler that moving artifacts between Github action jobs. Secondarily, personal perspective, it is simpler to maintain a single R script that does all the job than maintaining separated yaml files.

File: .github/workflows/package_building.yml (hunk 1 above)
Comment (zsigmas) [reply .github/workflows/package_building.yml:46]: Removed for debugging and creating the documentation, I will replace it

File: .github/workflows/package_building.yml (hunk 3 above)
Comment (zsigmas) [reply .github/workflows/package_building.yml:34]: They are now installed inside single build, but if we prefer a separate step for installing dependencies, no opposition.

File: pkg_pipeline/single_build.R (hunk 4 above)
Comment (zsigmas) [reply pkg_pipeline/single_build.R:20]: The qc report is used inside the test files to link specs and test cases, therefore it is required when running tests. In a next iteration we can think about ways to move it to dv.templates. No problem with that.

File: pkg_pipeline/single_build.R (hunk 5 above)
Comment (zsigmas) [reply pkg_pipeline/single_build.R:33]: I would argue that it is simpler to maintain a single R file that runs all the steps and offers a final summary with all the errors found. You don't wait more than having separate pods or have less information from my point of view, a table is included in the Action Summary, and in the end you have to check the logs anyway to find the specific errors.

File: pkg_pipeline/single_build.R (hunk 4 above)
Comment (sorinvoicu) [reply pkg_pipeline/single_build.R:20]: why not do it in the first place?

File: pkg_pipeline/single_build.R (hunk 5 above)
Comment (sorinvoicu) [reply pkg_pipeline/single_build.R:33]: old wheel to a new cart. disagree

File: .github/workflows/check.yml
Comment (sorinvoicu) [reply .github/workflows/check.yml:1]: I would still separate the concerns. I can chip in on the artifacts part if you want.

File: workflows.md (hunk 2 above)
Comment (zsigmas) [reply workflows.md:61]: solved in 8655fb0f251729d017ec1bf9c8eb6b72f3246d2c

File: .github/workflows/check.yml
Comment (sorinvoicu) [reply .github/workflows/check.yml:1]: Ideally we should just plug-and-play other qc reports/validatoR as the industry evolves. This ensures that.

File: pkg_pipeline/single_build.R (hunk 5 above)
Comment (ml-ebs-ext) [reply pkg_pipeline/single_build.R:32]: @sorinvoicu
Strongly agree with @zsigmas on this one and I'm ready to die on this hill.

"One cart carrying all the things we need VS several carts that need coordination."

Comment (sorinvoicu) [reply pkg_pipeline/single_build.R:32]: be my guest

Comment (ml-ebs-ext) [reply pkg_pipeline/single_build.R:32]: I have slept on the matter and have a few comments.

My combative wording ("ready to die on this hill") stems from your unfortunate "KISS, remember?" except I'm not insulting (accidentally or purposefully) anyone. In retrospect, I could have been more civil.

On the topic of simplicity, yes, I do think one big file that executes top-to-bottom is simpler than several files strung together through github actions. It is simpler because (in subjective decreasing order of importance):
1. You only need to look in a single place to understand how the build process works.
2. Related to the previous point, all build information is available at the same time on a single pod, making its manipulation trivial.
3. You can run it locally without having to mirror github machinery on your machine (a luxury that our internal colleagues don't have). This means you can tweak it locally and have a pretty good sense that it's going to work without having to go through minutes-long trial and error, which not only makes us miserable but costs the company money.
4. We have less work to do whenever github decides to deprecate features we rely on (https://github.blog/changelog/label/actions+deprecation/)
5. It makes it easier for potential users to take our code and use it mostly as-is outside of github. Our choice of hosting platform should not dictate theirs.
6. It is more environmentally friendly, because we save the cost of spinning up several containers (drop in the ocean, I know, considering we're encouraging the use of R).

I have numbered them so that you can comment on any of them individually and see where the nature of our disagreement lies. I would also like to know your reasons for thinking that a split `.R` file is simpler.

On the topic of my username: I revised it from `mlebs` to the obviously superior `ml-ebs-ext`, that make it clear that I'm an external working through EBS and my initials are "m" and "l".

Comment (ml-ebs-ext) [reply pkg_pipeline/single_build.R:32]: @sorinvoicu Sorry, I failed to mention your nickname yesterday. This is an important issue because it sits upstream of developers and dictates the way they work. I would like to hear your reasons for thinking that a split `.R` file is simpler. Even if we fail to convince each other, the whole team can benefit from this conversation.
